spotbugs integration: activate the 'run on full build' flag

Hojoe42 opened the pull request:
resolves #375

Hojoe42 committed:
run spotbugs for the entire project

Hojoe42 committed:
Merge branch 'm2e-code-quality:develop' into develop

Hojoe42 committed:
Merge branch 'm2e-code-quality:develop' into develop

adangel commented:
Thanks for the PR!
However, with your change, the test case com.basistech.m2e.code.quality.spotbugs.tests.EclipseSpotbugsProjectConfigurationTest.testSpotbugsReconfigureSkip is failing? Can you have a look at it?

Hojoe42 commented:
I cannot reproduce the error. I installed a new Eclipse 2025-12 and imported the ‘m2e-code-quality’ project as described in the README. The spotbugs tests run without errors in Eclipse and on the console. I am developing under Windows.
However, I have a test error in the PMD project. But I haven't changed anything there. The error occurred before I made any changes.

```
Results:

Failures:
  EclipsePmdProjectConfigurationTest.testPmdCheck:45->AbstractMavenProjectConfiguratorTestCase.importProjectRunBuildAndFindMarkers:75->AbstractMavenProjectConfiguratorTestCase.importProjectRunBuildAndFindMarkers:82->AbstractMavenProjectConfiguratorTestCase.runBuildAndAssertMarkers:94->assertMarkers:166 expected:<0> but was:<1>
  EclipsePmdProjectConfigurationTest.testPmdReconfigureSkip:79->assertMarkers:166 expected:<0> but was:<1>

Tests run: 7, Failures: 2, Errors: 0, Skipped: 0


!ENTRY org.eclipse.core.resources 2 10035 2026-01-02 13:58:19.049
!MESSAGE The workspace will exit with unsaved changes in this session.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for m2e-code-quality-plugins 2.3.1-SNAPSHOT:
[INFO]
[INFO] m2e-code-quality-plugins ........................... SUCCESS [  0.085 s]
[INFO] Shared Code for M2 Code Quality Plugin Extensions .. SUCCESS [  7.993 s]
[INFO] M2Eclipse Project Configurator for Eclipse Shared Tests SUCCESS [  0.886 s]
[INFO] M2Eclipse Project Configurator for Eclipse Checkstyle SUCCESS [  0.761 s]
[INFO] M2Eclipse Project Configurator for Eclipse Checkstyle Feature SUCCESS [  0.148 s]
[INFO] M2Eclipse Project Configurator for Eclipse Checkstyle Tests SUCCESS [ 28.223 s]
[INFO] M2Eclipse Project Configurator for Eclipse FindBugs  SUCCESS [  1.297 s]
[INFO] M2Eclipse Project Configurator for Eclipse FindBugs Feature SUCCESS [  0.154 s]
[INFO] M2Eclipse Project Configurator for Eclipse FindBugs Tests SUCCESS [ 21.947 s]
[INFO] M2Eclipse Project Configurator for Eclipse SpotBugs  SUCCESS [  1.259 s]
[INFO] M2Eclipse Project Configurator for Eclipse SpotBugs Feature SUCCESS [  0.155 s]
[INFO] M2Eclipse Project Configurator for Eclipse SpotBugs Tests SUCCESS [ 28.791 s]
[INFO] M2Eclipse Project Configurator for Eclipse PMD ..... SUCCESS [  2.564 s]
[INFO] M2Eclipse Project Configurator for Eclipse PMD Feature SUCCESS [  0.139 s]
[INFO] M2Eclipse Project Configurator for Eclipse PMD Tests FAILURE [ 45.377 s]
[INFO] M2Eclipse Code Quality Plugins Update Site ......... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:23 min
[INFO] Finished at: 2026-01-02T13:58:19+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-surefire-plugin:5.0.1:test (default-test) on project com.basistech.m2e.code.quality.pmd.test: There are test failures.
[ERROR]
[ERROR] Please refer to C:\programmierung\Eclipse_Installer\m2e-code-quality-develop\ws\m2e-code-quality\com.basistech.m2e.code.quality.pmd.test\target\surefire-reports for the individual test results.`
```

Hojoe42 committed:
Merge branch 'm2e-code-quality:develop' into develop

adangel committed:
spotbugs (test): Do not trigger spotbugs explicitly anymore

Spotbugs should be executed (as the other plugins) during a build
via its ProjectBuilder.

adangel committed:
Merge branch 'develop' into pr-402

adangel reviewed (approved):
I've now checked the code and I came to the conclusion, that disabling spotbugs for full builds (the line you removed) is actually fixing a bug:

- As you mentioned, this flag is enabled by default - so, Spotbugs should actually run by default for a full build. I don't see a reason, why it shouldn't.
- The tests for Spotbugs always executed Spotbugs explicitly, so they didn't actually verify the project builders. For the other tools (checkstyle, pmd), the test projects are just built and markers are expected. I've adjusted the Spotbugs test now to be the same (almost - needed to a WaitForFindbugsJob).
- A reason maybe, why Spotbugs was triggered explicitly: Spotbugs runs the analysis in a separate job by default - which means, the markers are not ready when the build is complete. It takes some time afterwards, until Spotbugs ran. I guess it was just easier at that time to explicitly trigger Spotbugs rather than do an extra wait.
I've added this extra wait now - I hope it's reliably enough.

For the Spotbugs test failures - I couldn't reproduce these locally either. A second run on Github Actions also made the test failure on the Spotbugs-Tests disappear. So, it's definitively a timing problem (looks like somehow Spotbugs analysis was triggered after all by a incremental build in the test case before the builders where removed and then the markers appeared when they shouldn't).

For the PMD test failures, I didn't see them locally. Let's see how the GitHub Action run works...
